## A branch-and-bound example (see slides Part II)

In [1]:
import gurobipy as gb

In [2]:
knap_lr = gb.read('BBExample.lp')

Academic license - for non-commercial use only - expires 2021-05-15
Using license file /Users/fabrizio/Solvers/licenses/gurobi.lic
Read LP format model from file BBExample.lp
Reading time = 0.00 seconds
: 1 rows, 7 columns, 7 nonzeros


$$ 
\max 9 x_1 + 15 x_2 + 8 x_3 + 6 x_4 + 5 x_5 + 4 x_6 + x_7 \\
\text{s. t.}\\
6 x_1 + 11 x_2 + 6 x_3 + 5 x_4 + 5 x_5 + 4 x_6 + x_7 \le 19\\
x \in \{0,1 \}^7
$$

### Branch-and-bound initialization

1. Subproblem list $L$ is initialized with the problem itself
2. $z^H$ = 0


### Start the loop

1. Pick a subproblem from the list
2. Solve the linear relaxation

In [3]:
knap_lr.optimize()

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Model fingerprint: 0x6ec019d8
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
Presolve removed 0 rows and 2 columns
Presolve time: 0.01s
Presolved: 1 rows, 5 columns, 5 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    2.8500000e+01   2.166667e+00   0.000000e+00      0s
       1    2.6666667e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.02 seconds
Optimal objective  2.666666667e+01


$$z^0_{UB} = 26.6666$$

In [42]:
solution_S0 = {x.VarName:x.x for x in knap_lr.getVars()}
S0_bound = knap_lr.objVal


print('z(S0):', S0_bound)
print (solution_S0)

z(S0): 26.666666666666668
{'x1': 1.0, 'x2': 1.0, 'x3': 0.3333333333333333, 'x4': 0.0, 'x5': 0.0, 'x6': 0.0, 'x7': 0.0}


### Branching on variable $x_3$

$$ S_1 = S_0 \cap \{x_3 = 0\} \\
S_2 = S_0 \cap \{x_3 = 1\}$$

$$L = \{S_1, S_2 \} $$

$$ z^1_{UB} \le 26.6666$$

$$ z^2_{UB} \le 26.6666$$



### Second iteration

1. Select a problem from $L$, say $S_1$
2. Explore node $S_1$
3. Solve the linear relaxation of $S_1$

### Subproblem $S_1$

In [4]:
knap_lr.getVarByName('x3').lb = 0
knap_lr.getVarByName('x3').ub = 0

knap_lr.optimize()

S1_bound = knap_lr.objVal


solution_S1 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S1):', S1_bound)
print(solution_S1)


Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.6666667e+01   1.666667e-01   0.000000e+00      0s
       1    2.6400000e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.02 seconds
Optimal objective  2.640000000e+01
z(S1): 26.4
{'x1': 1.0, 'x2': 1.0, 'x3': 0.0, 'x4': 0.4, 'x5': 0.0, 'x6': 0.0, 'x7': 0.0}


#### Branching on variable $x_4$

$$ S_3 = S_1 \cap \{x_4 = 0\} \\
S_4 = S_1 \cap \{x_4 = 1\}$$

$$L = \{S_2, S_3, S_4 \} $$

$$ z^3_{UB} \le 26.4$$

$$ z^4_{UB} \le 26.4$$




### Subproblem $S_2$

In [5]:
knap_lr.getVarByName('x3').lb = 1
knap_lr.getVarByName('x3').ub = 1

knap_lr.optimize()

S2_bound = knap_lr.objVal


solution_S2 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S2):', S2_bound)
print(solution_S2)

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.7200000e+01   4.000000e-01   0.000000e+00      0s
       1    2.6545455e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.02 seconds
Optimal objective  2.654545455e+01
z(S2): 26.545454545454547
{'x1': 1.0, 'x2': 0.6363636363636364, 'x3': 1.0, 'x4': 0.0, 'x5': 0.0, 'x6': 0.0, 'x7': 0.0}


#### Branching on variable $x_2$

$$ S_5 = S_2 \cap \{x_2 = 0\} \\
S_6 = S_2 \cap \{x_2 = 1\}$$

$$L = \{S_3, S_4, S_5, S_6 \} $$

$$ z^5_{UB} \le 26.54$$

$$ z^6_{UB} \le 26.54$$





### Subproblem $S_5$

In [6]:
knap_lr.getVarByName('x3').lb = 1
knap_lr.getVarByName('x3').ub = 1

knap_lr.getVarByName('x2').lb = 0
knap_lr.getVarByName('x2').ub = 0

knap_lr.optimize()

S5_bound = knap_lr.objVal


solution_S5 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S5):', S5_bound)
print(solution_S5)

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.6545455e+01   6.363636e-01   0.000000e+00      0s
       2    2.5000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.01 seconds
Optimal objective  2.500000000e+01
z(S5): 25.0
{'x1': 1.0, 'x2': 0.0, 'x3': 1.0, 'x4': 1.0, 'x5': 0.0, 'x6': 0.25, 'x7': 1.0}


#### Branching on variable $x_6$

$$ S_7 = S_5 \cap \{x_6 = 0\} \\
S_8 = S_5 \cap \{x_6 = 1\}$$

$$L = \{S_3, S_4, S_6, S_7, S_8 \} $$

$$ z^7_{UB} \le 25$$

$$ z^8_{UB} \le 25$$




### Subproblem $S_6$

In [7]:
knap_lr.getVarByName('x3').lb = 1
knap_lr.getVarByName('x3').ub = 1

knap_lr.getVarByName('x2').lb = 1
knap_lr.getVarByName('x2').ub = 1

knap_lr.optimize()

S6_bound = knap_lr.objVal


solution_S6 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S6):', S6_bound)
print(solution_S6)

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.9000000e+01   6.250000e-01   0.000000e+00      0s
       1    2.6000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.02 seconds
Optimal objective  2.600000000e+01
z(S6): 26.0
{'x1': 0.3333333333333333, 'x2': 1.0, 'x3': 1.0, 'x4': 0.0, 'x5': 0.0, 'x6': 0.0, 'x7': 0.0}


### Branching on variable $x_1$

$$ S_9 = S_6 \cap \{x_1 = 0\} \\
S_{10} = S_6 \cap \{x_1 = 1\}$$

$$L = \{S_3, S_4, S_7, S_8, S_9, S_{10} \} $$

$$ z^9_{UB} \le 26$$

$$ z^{10}_{UB} \le 26$$

### Subproblem $S_3$

In [8]:
for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 1.0 1.0
x3 1.0 1.0
x4 0.0 1.0
x5 0.0 1.0
x6 0.0 1.0
x7 0.0 1.0


In [10]:
knap_lr.getVarByName('x3').lb = 0
knap_lr.getVarByName('x3').ub = 0

knap_lr.getVarByName('x2').lb = 0
knap_lr.getVarByName('x2').ub = 1

knap_lr.getVarByName('x4').lb = 0
knap_lr.getVarByName('x4').ub = 0

knap_lr.update()

for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 0.0 0.0
x5 0.0 1.0
x6 0.0 1.0
x7 0.0 1.0


In [11]:
knap_lr.optimize()

S3_bound = knap_lr.objVal


solution_S3 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S3):', S3_bound)
print(solution_S3)

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.8500000e+01   1.083333e+00   0.000000e+00      0s
       1    2.6000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.01 seconds
Optimal objective  2.600000000e+01
z(S3): 26.0
{'x1': 1.0, 'x2': 1.0, 'x3': 0.0, 'x4': 0.0, 'x5': 0.0, 'x6': 0.25, 'x7': 1.0}


### Branching on variable $x_6$

$$ S_{11} = S_3 \cap \{x_6 = 0\} \\
S_{12} = S_3 \cap \{x_6 = 0\}$$

$$L = \{S_4, S_7, S_8, S_9, S_{10}, S_{11}, S_{12} \} $$

$$ z^{11}_{UB} \le 26$$

$$ z^{12}_{UB} \le 26$$

In [12]:
for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 0.0 0.0
x5 0.0 1.0
x6 0.0 1.0
x7 0.0 1.0


### Subproblem $S_4$

In [13]:
knap_lr.getVarByName('x4').lb = 1
knap_lr.getVarByName('x4').ub = 1

knap_lr.update()

for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 1.0 1.0
x5 0.0 1.0
x6 0.0 1.0
x7 0.0 1.0


In [14]:
knap_lr.optimize()

S4_bound = knap_lr.objVal


solution_S4 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S4):', S4_bound)
print(solution_S4)

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.7000000e+01   2.500000e-01   0.000000e+00      0s
       1    2.5909091e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.01 seconds
Optimal objective  2.590909091e+01
z(S4): 25.90909090909091
{'x1': 1.0, 'x2': 0.7272727272727273, 'x3': 0.0, 'x4': 1.0, 'x5': 0.0, 'x6': 0.0, 'x7': 0.0}


### Branching on variable $x_2$

$$ S_{13} = S_4 \cap \{x_2 = 0\} \\
S_{14} = S_4 \cap \{x_2 = 1\}$$

$$L = \{S_7, S_8, S_9, S_{10}, S_{11}, S_{12}, S_{13}, S_{14} \} $$

$$ z^{13}_{UB} \le 25.909$$

$$ z^{14}_{UB} \le 25.909$$

### Subproblem $S_{11}$

In [15]:
knap_lr.getVarByName('x4').lb = 0
knap_lr.getVarByName('x4').ub = 0

knap_lr.getVarByName('x6').lb = 0
knap_lr.getVarByName('x6').ub = 0

knap_lr.update()

for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 0.0 0.0
x5 0.0 1.0
x6 0.0 0.0
x7 0.0 1.0


In [16]:
knap_lr.optimize()

S11_bound = knap_lr.objVal


solution_S11 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S11):', S11_bound)
print(solution_S11)

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.6727273e+01   1.818182e-01   0.000000e+00      0s
       2    2.6000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.02 seconds
Optimal objective  2.600000000e+01
z(S11): 26.0
{'x1': 1.0, 'x2': 1.0, 'x3': 0.0, 'x4': 0.0, 'x5': 0.2, 'x6': 0.0, 'x7': 1.0}


### Branching on variable $x_5$

$$ S_{15} = S_{11} \cap \{x_5 = 0\} \\
S_{16} = S_{11} \cap \{x_5 = 1\}$$

$$L = \{S_7, S_8, S_9, S_{10},  S_{12}, S_{13}, S_{14}, S_{15}, S_{16} \} $$

$$ z^{15}_{UB} \le 26$$

$$ z^{16}_{UB} \le 26$$

### Subproblem $S_{15}$

In [17]:
knap_lr.getVarByName('x5').lb = 0
knap_lr.getVarByName('x5').ub = 0


knap_lr.update()

for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 0.0 0.0
x5 0.0 0.0
x6 0.0 0.0
x7 0.0 1.0


In [18]:
knap_lr.optimize()

S15_bound = knap_lr.objVal


solution_S15 = {x.VarName:x.x for x in knap_lr.getVars()}

print('z(S15):', S15_bound)
print(solution_S15)

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.6000000e+01   1.000000e-01   0.000000e+00      0s
       1    2.5000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.01 seconds
Optimal objective  2.500000000e+01
z(S15): 25.0
{'x1': 1.0, 'x2': 1.0, 'x3': 0.0, 'x4': 0.0, 'x5': 0.0, 'x6': 0.0, 'x7': 1.0}


After incumbent's update one can remove from the list 
subproblems $S_{13}, S_{14}, S_{7}, S_{8}$

$$L = \{S_9, S_{10},  S_{12}, S_{16} \} $$

### Subproblem $S_{16}$

In [19]:
knap_lr.getVarByName('x5').lb = 1
knap_lr.getVarByName('x5').ub = 1


knap_lr.update()

for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 0.0 0.0
x5 1.0 1.0
x6 0.0 0.0
x7 0.0 1.0


In [20]:
knap_lr.optimize()

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    3.0000000e+01   2.500000e-01   0.000000e+00      0s
       1    2.4909091e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.04 seconds
Optimal objective  2.490909091e+01


$$L = \{S_9, S_{10},  S_{12} \} $$

### Subproblem $S_{12}$


In [22]:
for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 0.0 0.0
x5 1.0 1.0
x6 0.0 0.0
x7 0.0 1.0


In [23]:
knap_lr.getVarByName('x5').lb = 0
knap_lr.getVarByName('x5').ub = 1

knap_lr.getVarByName('x6').lb = 1
knap_lr.getVarByName('x6').ub = 1

knap_lr.optimize()

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.5272727e+01   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.01 seconds
Optimal objective  2.527272727e+01


$$L = \{S_9, S_{10}\} $$

### Subproblem $S_{9}$

In [24]:
for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 0.0 1.0
x2 0.0 1.0
x3 0.0 0.0
x4 0.0 0.0
x5 0.0 1.0
x6 1.0 1.0
x7 0.0 1.0


In [25]:
knap_lr.getVarByName('x4').lb = 0
knap_lr.getVarByName('x4').ub = 1

knap_lr.getVarByName('x6').lb = 0
knap_lr.getVarByName('x6').ub = 1

knap_lr.getVarByName('x3').lb = 1
knap_lr.getVarByName('x3').ub = 1

knap_lr.getVarByName('x2').lb = 1
knap_lr.getVarByName('x2').ub = 1

knap_lr.getVarByName('x1').lb = 0
knap_lr.getVarByName('x1').ub = 0

knap_lr.update()

for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)



x1 0.0 0.0
x2 1.0 1.0
x3 1.0 1.0
x4 0.0 1.0
x5 0.0 1.0
x6 0.0 1.0
x7 0.0 1.0


In [26]:
knap_lr.optimize()

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.5727273e+01   1.818182e-01   0.000000e+00      0s
       1    2.5400000e+01   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.01 seconds
Optimal objective  2.540000000e+01


$$L = \{S_{10}\} $$

### Subproblem $S_{10}$

In [28]:
knap_lr.getVarByName('x1').lb = 1
knap_lr.getVarByName('x1').ub = 1

knap_lr.update()

for x in knap_lr.getVars():
    print (x.VarName, x.lb, x.ub)

x1 1.0 1.0
x2 1.0 1.0
x3 1.0 1.0
x4 0.0 1.0
x5 0.0 1.0
x6 0.0 1.0
x7 0.0 1.0


In [29]:
knap_lr.optimize()

Gurobi Optimizer version 9.1.1 build v9.1.1rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 7 columns and 7 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 1e+01]
  Objective range  [1e+00, 2e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [2e+01, 2e+01]
       0    2.7200000e+01   4.000000e-01   0.000000e+00      0s

Solved in 0 iterations and 0.02 seconds
Infeasible model


$$L = \emptyset$$